### PR TITLE
MINIFICPP-1494 Allow InvokeHTTP GET requests without incoming flowfile

### PR DESCRIPTION
**Pull request by lordgamez**
Handling the use case when InvokeHTTP processor has no real incoming connection but there is a Retry self-loop, the scheduler should trigger the processor.

Jira ticket: https://issues.apache.org/jira/browse/MINIFICPP-1494

Note: We discussed that renaming the relationships is not an option to avoid introducing a breaking change, only change (1) is implemented

---------------------------------------------------------------------------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: PROCESSORS.md [hunk 1]
```diff
@@ -704,6 +704,10 @@ In the list below, the names of required properties appear in bold. Any other pr
 | Name | Description |
 | - | - |
 |success|All files are routed to success|
```
Comment (fgerlits) [inline PROCESSORS.md:706]: I think this could be clarified, eg. `All incoming FlowFiles are routed, unchanged, to succes`

File: extensions/librdkafka/tests/ConsumeKafkaTests.cpp [hunk 2]
```diff
@@ -422,7 +422,7 @@ class ConsumeKafkaContinuousPublishingTest : public ConsumeKafkaTest {
       int64_t value;
       org::apache::nifi::minifi::core::TimeUnit unit;
       REQUIRE(org::apache::nifi::minifi::core::Property::StringToTime(property_string, value, unit));
-      int64_t value_as_ms;
+      int64_t value_as_ms = 0;
```
Comment (fgerlits) [inline extensions/librdkafka/tests/ConsumeKafkaTests.cpp:425]: this should be in a different pull request, as it is not related to this change

File: PROCESSORS.md (hunk 1 above)
Comment (lordgamez) [reply PROCESSORS.md]: This is not true, I updated it in ddb119bc9a787d6d5c8600dfd2c7ed22e72da1ed to be on par with the NiFi description.

File: extensions/http-curl/tests/HttpGetIntegrationTest.cpp [hunk 3]
```diff
@@ -63,78 +62,134 @@ class HttpResponder : public CivetHandler {
...
-  LogTestController::getInstance().setDebug<minifi::processors::LogAttribute>();
+class VerifyHTTPGet {
+ public:
+  VerifyHTTPGet(const cmd_args& args, CivetHandler& http_handler)
+      : args_(args)
+      , configuration_(std::make_shared<minifi::Configure>())
+      , test_repo_(std::make_shared<TestRepository>())
+      , test_flow_repo_(std::make_shared<TestFlowRepository>())
+      , content_repo_(std::make_shared<core::repository::VolatileContentRepository>())
+      , stream_factory_(minifi::io::StreamFactory::getInstance(configuration_)) {
```
Comment (martinzink) [inline extensions/http-curl/tests/HttpGetIntegrationTest.cpp:82]: very minor, but I feel like the comma should be in the end of the previous lines, to comply with our style guide
https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists

Comment (lordgamez) [reply extensions/http-curl/tests/HttpGetIntegrationTest.cpp]: You are right, that was just some ad-hoc style I was used to. Fixed it in a0f1f41e01915b75f2c3f4ca15162fae9fb587a6

File: extensions/http-curl/tests/HttpGetIntegrationTest.cpp [hunk 4]
```diff
@@ -63,78 +62,134 @@ class HttpResponder : public CivetHandler {
...
-  core::YamlConfiguration yaml_config(test_repo, test_repo, content_repo, stream_factory, configuration, args.test_file);
+    const auto inv = std::dynamic_pointer_cast<minifi::processors::InvokeHTTP>(proc);
+    assert(inv != nullptr);
 
-  std::shared_ptr<core::Processor> proc = yaml_config.getRoot()->findProcessorByName("invoke");
-  assert(proc != nullptr);
+    inv->getProperty(minifi::processors::InvokeHTTP::URL.getName(), url_);
+  }
+
+  void setupServer(CivetHandler& http_handler) {
```
Comment (adamdebreceni) [inline extensions/http-curl/tests/HttpGetIntegrationTest.cpp:139]: there seems to be a similar class `VerifyRetryInvokeHTTP` in `VerifyInvokeHTTPTests.cpp`, how do the two relate?

File: extensions/http-curl/tests/HttpGetIntegrationTest.cpp [hunk 5]
```diff
@@ -63,78 +62,134 @@ class HttpResponder : public CivetHandler {
...
+  }
+};
 
-  LogTestController::getInstance().setDebug<core::Processor>();
-  LogTestController::getInstance().setDebug<core::ProcessSession>();
-  LogTestController::getInstance().setDebug<utils::HTTPClient>();
-  LogTestController::getInstance().setDebug<minifi::controllers::SSLContextService>();
-  LogTestController::getInstance().setDebug<minifi::processors::InvokeHTTP>();
-  LogTestController::getInstance().setDebug<minifi::processors::LogAttribute>();
+class VerifyHTTPGet {
```
Comment (adamdebreceni) [inline extensions/http-curl/tests/HttpGetIntegrationTest.cpp:74]: there seems to be a similar class `VerifyRetryInvokeHTTP` in `VerifyInvokeHTTPTests.cpp`, how do the two relate?

Comment (lordgamez) [reply extensions/http-curl/tests/HttpGetIntegrationTest.cpp:74]: VerifyInvokeHTTPTests is similar in design, but it has verifications with callbacks for POST requests and only uses yaml configuration files with POST invoke operations.

Comment (adamdebreceni) [reply extensions/http-curl/tests/HttpGetIntegrationTest.cpp:74]: could we somehow reuse it (or any of its base classes `VeifyInvokeHTTP`, `HTTPIntegrationBase`) as base class of `VerifyHTTPGet`? they provide some of functionality implemented here (setting up the server, possibly the flow itself)

Comment (lordgamez) [reply extensions/http-curl/tests/HttpGetIntegrationTest.cpp:74]: I wanted to keep it separate as the `HTTPIntegrationBase` has way more complexity than we need, but I'll check if we could reuse it and reduce any duplications.

Comment (lordgamez) [reply extensions/http-curl/tests/HttpGetIntegrationTest.cpp]: I did some refactoring and reused the `VeifyInvokeHTTP` class for the `VerifyHTTPGet` in 13b46dc6e39687ec63bd545127bae8a6012c7327

File: extensions/http-curl/tests/VerifyInvokeHTTP.h [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    if (path_) {
+      throw std::logic_error("Url is already set");
+    }
+    std::string port, scheme, path;
+    parse_http_components(url, port, scheme, path);
+    path_ = path;
+    HTTPIntegrationBase::setUrl(url, handler);
+  }
+
+  void setProperties(std::shared_ptr<core::Processor> proc) {
```
Comment (adam-markovics) [inline extensions/http-curl/tests/VerifyInvokeHTTP.h:53]: Is it necessary to increase the ref count of proc?

File: extensions/http-curl/tests/VerifyInvokeHTTP.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    std::string url = scheme + "://localhost:" + getWebPort() + *path_;
+    proc->setProperty(minifi::processors::InvokeHTTP::URL.getName(), url);
+  }
+
+  void setProperty(const std::string& property, const std::string& value) {
+    const auto components = flowController_->getComponents("InvokeHTTP");
+    assert(!components.empty());
+
+    const auto stateController = components.at(0);
```
Comment (adam-markovics) [inline extensions/http-curl/tests/VerifyInvokeHTTP.h:62]: This can even be accessed using [], since emptiness is checked before.

File: extensions/http-curl/tests/VerifyInvokeHTTP.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    std::shared_ptr<core::Repository> test_flow_repo = std::make_shared<TestFlowRepository>();
+
+    if (flow_yml_path) {
+      configuration->set(minifi::Configure::nifi_flow_configuration_file, *flow_yml_path);
+    }
+    configuration->set("c2.agent.heartbeat.period", "200");
+    std::shared_ptr<core::ContentRepository> content_repo = std::make_shared<core::repository::VolatileContentRepository>();
+    content_repo->initialize(configuration);
+    std::shared_ptr<minifi::io::StreamFactory> stream_factory = minifi::io::StreamFactory::getInstance(configuration);
+    std::unique_ptr<core::FlowConfiguration> yaml_ptr = std::unique_ptr<core::YamlConfiguration>(
```
Comment (adam-markovics) [inline extensions/http-curl/tests/VerifyInvokeHTTP.h:83]: Could be make_unique.

File: extensions/http-curl/tests/VerifyInvokeHTTP.h (hunk 6 above)
Comment (lordgamez) [reply extensions/http-curl/tests/VerifyInvokeHTTP.h]: Updated in 7f194ba73816ee7fb8464089813e29573106e506

File: extensions/http-curl/tests/VerifyInvokeHTTP.h (hunk 7 above)
Comment (lordgamez) [reply extensions/http-curl/tests/VerifyInvokeHTTP.h]: Updated in 7f194ba73816ee7fb8464089813e29573106e506

File: extensions/http-curl/tests/VerifyInvokeHTTP.h (hunk 8 above)
Comment (lordgamez) [reply extensions/http-curl/tests/VerifyInvokeHTTP.h]: Updated in 7f194ba73816ee7fb8464089813e29573106e506

**Review by adam-markovics — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]

File: extensions/http-curl/processors/InvokeHTTP.cpp [hunk 9]
```diff
@@ -123,9 +123,12 @@ const char* InvokeHTTP::REMOTE_DN = "invokehttp.remote.dn";
 const char* InvokeHTTP::EXCEPTION_CLASS = "invokehttp.java.exception.class";
 const char* InvokeHTTP::EXCEPTION_MESSAGE = "invokehttp.java.exception.message";
 
-core::Relationship InvokeHTTP::Success("success", "All files are routed to success");
+core::Relationship InvokeHTTP::Success("success", "The original FlowFile will be routed upon success (2xx status codes). "
+                                       "It will have new attributes detailing the success of the request.");
 
-core::Relationship InvokeHTTP::RelResponse("response", "Represents a response flowfile");
+core::Relationship InvokeHTTP::RelResponse("response", "A Response FlowFile will be routed upon success (2xx status codes). "
+                                           "If the 'Output Response Regardless' property is true then the response will be sent "
```
Comment (szaszm) [inline extensions/http-curl/processors/InvokeHTTP.cpp:130]:
```suggestion
                                           "If the 'Always Output Response' property is true then the response will be sent "
```

Comment (lordgamez) [reply extensions/http-curl/processors/InvokeHTTP.cpp]: Good catch! Fixed in 1445eb3bcdf9367440314d513718e0afa23ee2ee

**Review by szaszm — APPROVED** [review]

**Review by adamdebreceni — APPROVED** [review]
